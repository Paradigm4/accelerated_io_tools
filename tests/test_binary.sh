#!/bin/bash

set -o errexit

DIR=`dirname $0`
TEST_OUT=$DIR/test_binary.out
IQ="iquery --afl --query"
F=/tmp/1

function iq() {
    $IQ "$1" >> $TEST_OUT
}

: > $TEST_OUT                   # Reset output file


echo -e "\nI. Basic"
echo "1. Single chunk (int64)"
iq "aio_save(apply(build(<x:int64>[i=-10:10:0:21], i), y, iif(i%2=0, i, int64(null))), '$F', 'atts_only=0', 'format=(int64 null,int64 null,int64)')"
iq "input(<x:int64, y:int64, i:int64 not null>[j], '$F', 0, '(int64 null,int64 null,int64)')"

echo "2. Multiple chunks (int64)"
iq "aio_save(apply(build(<x:int64>[i=1:100:0:20], i), y, iif(i%2=0, i, int64(null))), '$F', 'atts_only=0', 'format=(int64 null,int64 null,int64)')"
iq "input(<x:int64, y:int64, i:int64 not null>[j], '$F', 0, '(int64 null,int64 null,int64)')"

echo "3. Multiple chunks, multiple dimensions (int64)"
iq "aio_save(
      apply(
        build(<x:int64>[i=1:30:0:10; j=1:10:0:5], i * j),
        y, iif(i%2=0, x, int64(null))),
      '$F',
      'atts_only=0',
      'format=(int64 null,int64 null,int64,int64)')"
iq "input(<x:int64, y:int64, i:int64 not null, j:int64 not null>[k], '$F', 0, '(int64 null,int64 null,int64,int64)')"

echo "4. bool"
iq "aio_save(apply(build(<x:bool not null>[i=1:20:0:4], i), y, iif(i%2=0, bool(i), bool(null))), '$F', 'atts_only=0', 'format=(bool,bool null,int64)')"
iq "input(<x:bool not null, y:bool, i:int64 not null>[j], '$F', 0, '(bool,bool null,int64)')"

echo "5. double"
iq "aio_save(
      apply(
        build(<x:double not null>[i=1:20:0:4], i / 10.),
        y, iif(i%2=0, double(i) / 10, double(null))),
      '$F',
      'atts_only=0',
      'format=(double,double null,int64)')"
iq "input(<x:double not null, y:double, i:int64 not null>[j], '$F', 0, '(double,double null,int64)')"

echo "6. float"
iq "aio_save(
      apply(
        build(<x:float not null>[i=1:20:0:4], i / 10.),
        y, iif(i%2=0, float(i) / 10, float(null))),
      '$F',
      'atts_only=0',
      'format=(float,float null,int64)')"
iq "input(<x:float not null, y:float, i:int64 not null>[j], '$F', 0, '(float,float null,int64)')"

echo "7. uint64"
iq "aio_save(
      apply(
        build(<x:uint64 not null>[i=1:20:0:4], i),
        y, iif(i%2=0, uint64(i), uint64(null))),
      '$F',
      'atts_only=0',
      'format=(uint64,uint64 null,int64)')"
iq "input(<x:uint64 not null, y:uint64, i:int64 not null>[j], '$F', 0, '(uint64,uint64 null,int64)')"

echo "8. int64, double, string"
iq "aio_save(
      apply(
        build(<x:int64>[i=1:20:0:4], i),
        y, iif(i%2=0, double(i) / 10, double(null)),
        z, iif(i%2=0, string(i), string(null))),
      '$F',
      'atts_only=0',
      'format=(int64 null,double null,string null,int64)')"
iq "input(<x:int64, y:double, z:string, i:int64 not null>[j], '$F', 0, '(int64 null,double null,string null,int64)')"

echo "9. Empty chunk"
iq "aio_save(filter(build(<x:int64>[i=1:20:0:4], i), x < 6 or x > 14), '$F', 'atts_only=0', 'format=(int64 null,int64)')"
iq "input(<x:int64, i:int64 not null>[j], '$F', 0, '(int64 null,int64)')"


echo -e "\nII. Exceptions"
echo "1. Incomplete format (missing)"
$IQ "aio_save(build(<x:int64 not null>[i=0:0], i), '$F', 'atts_only=0', 'format=(int64)')" 2>&1 \
    |  sed --expression='s/ line: [0-9]\+//g'                                                   \
    |  grep --invert-match "Failed query id:" >> $TEST_OUT                                      \
    || echo "expected exception"


echo "2. Incomplete format (extra)"
$IQ "aio_save(build(<x:int64 not null>[i=0:0], i), '$F', 'atts_only=0', 'format=(int64,int64,int64)')" 2>&1     \
    |  sed --expression='s/ line: [0-9]\+//g'                                                                   \
    |  grep --invert-match "Failed query id:" >> $TEST_OUT                                                      \
    || echo "expected exception"


echo "3. Incomplete format (wrong)"
$IQ "aio_save(build(<x:int64 not null>[i=0:0], i), '$F', 'atts_only=0', 'format=(int64,double)')" 2>&1  \
    |  sed --expression='s/ line: [0-9]\+//g'                                                           \
    |  grep --invert-match "Failed query id:" >> $TEST_OUT                                              \
    || echo "expected exception"


diff $TEST_OUT $DIR/test_binary.expected
