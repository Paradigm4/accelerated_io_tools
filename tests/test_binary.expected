{chunk_no,dest_instance_id,source_instance_id} val
{j} x,y,i
{0} -10,-10,-10
{1} -9,null,-9
{2} -8,-8,-8
{3} -7,null,-7
{4} -6,-6,-6
{5} -5,null,-5
{6} -4,-4,-4
{7} -3,null,-3
{8} -2,-2,-2
{9} -1,null,-1
{10} 0,0,0
{11} 1,null,1
{12} 2,2,2
{13} 3,null,3
{14} 4,4,4
{15} 5,null,5
{16} 6,6,6
{17} 7,null,7
{18} 8,8,8
{19} 9,null,9
{20} 10,10,10
{chunk_no,dest_instance_id,source_instance_id} val
{j} x,y,i
{0} 21,null,21
{1} 22,22,22
{2} 23,null,23
{3} 24,24,24
{4} 25,null,25
{5} 26,26,26
{6} 27,null,27
{7} 28,28,28
{8} 29,null,29
{9} 30,30,30
{10} 31,null,31
{11} 32,32,32
{12} 33,null,33
{13} 34,34,34
{14} 35,null,35
{15} 36,36,36
{16} 37,null,37
{17} 38,38,38
{18} 39,null,39
{19} 40,40,40
{20} 61,null,61
{21} 62,62,62
{22} 63,null,63
{23} 64,64,64
{24} 65,null,65
{25} 66,66,66
{26} 67,null,67
{27} 68,68,68
{28} 69,null,69
{29} 70,70,70
{30} 71,null,71
{31} 72,72,72
{32} 73,null,73
{33} 74,74,74
{34} 75,null,75
{35} 76,76,76
{36} 77,null,77
{37} 78,78,78
{38} 79,null,79
{39} 80,80,80
{40} 81,null,81
{41} 82,82,82
{42} 83,null,83
{43} 84,84,84
{44} 85,null,85
{45} 86,86,86
{46} 87,null,87
{47} 88,88,88
{48} 89,null,89
{49} 90,90,90
{50} 91,null,91
{51} 92,92,92
{52} 93,null,93
{53} 94,94,94
{54} 95,null,95
{55} 96,96,96
{56} 97,null,97
{57} 98,98,98
{58} 99,null,99
{59} 100,100,100
{60} 1,null,1
{61} 2,2,2
{62} 3,null,3
{63} 4,4,4
{64} 5,null,5
{65} 6,6,6
{66} 7,null,7
{67} 8,8,8
{68} 9,null,9
{69} 10,10,10
{70} 11,null,11
{71} 12,12,12
{72} 13,null,13
{73} 14,14,14
{74} 15,null,15
{75} 16,16,16
{76} 17,null,17
{77} 18,18,18
{78} 19,null,19
{79} 20,20,20
{80} 41,null,41
{81} 42,42,42
{82} 43,null,43
{83} 44,44,44
{84} 45,null,45
{85} 46,46,46
{86} 47,null,47
{87} 48,48,48
{88} 49,null,49
{89} 50,50,50
{90} 51,null,51
{91} 52,52,52
{92} 53,null,53
{93} 54,54,54
{94} 55,null,55
{95} 56,56,56
{96} 57,null,57
{97} 58,58,58
{98} 59,null,59
{99} 60,60,60
{chunk_no,dest_instance_id,source_instance_id} val
{k} x,y,i,j
{0} 11,null,11,1
{1} 22,null,11,2
{2} 33,null,11,3
{3} 44,null,11,4
{4} 55,null,11,5
{5} 12,12,12,1
{6} 24,24,12,2
{7} 36,36,12,3
{8} 48,48,12,4
{9} 60,60,12,5
{10} 13,null,13,1
{11} 26,null,13,2
{12} 39,null,13,3
{13} 52,null,13,4
{14} 65,null,13,5
{15} 14,14,14,1
{16} 28,28,14,2
{17} 42,42,14,3
{18} 56,56,14,4
{19} 70,70,14,5
{20} 15,null,15,1
{21} 30,null,15,2
{22} 45,null,15,3
{23} 60,null,15,4
{24} 75,null,15,5
{25} 16,16,16,1
{26} 32,32,16,2
{27} 48,48,16,3
{28} 64,64,16,4
{29} 80,80,16,5
{30} 17,null,17,1
{31} 34,null,17,2
{32} 51,null,17,3
{33} 68,null,17,4
{34} 85,null,17,5
{35} 18,18,18,1
{36} 36,36,18,2
{37} 54,54,18,3
{38} 72,72,18,4
{39} 90,90,18,5
{40} 19,null,19,1
{41} 38,null,19,2
{42} 57,null,19,3
{43} 76,null,19,4
{44} 95,null,19,5
{45} 20,20,20,1
{46} 40,40,20,2
{47} 60,60,20,3
{48} 80,80,20,4
{49} 100,100,20,5
{50} 1,null,1,1
{51} 2,null,1,2
{52} 3,null,1,3
{53} 4,null,1,4
{54} 5,null,1,5
{55} 2,2,2,1
{56} 4,4,2,2
{57} 6,6,2,3
{58} 8,8,2,4
{59} 10,10,2,5
{60} 3,null,3,1
{61} 6,null,3,2
{62} 9,null,3,3
{63} 12,null,3,4
{64} 15,null,3,5
{65} 4,4,4,1
{66} 8,8,4,2
{67} 12,12,4,3
{68} 16,16,4,4
{69} 20,20,4,5
{70} 5,null,5,1
{71} 10,null,5,2
{72} 15,null,5,3
{73} 20,null,5,4
{74} 25,null,5,5
{75} 6,6,6,1
{76} 12,12,6,2
{77} 18,18,6,3
{78} 24,24,6,4
{79} 30,30,6,5
{80} 7,null,7,1
{81} 14,null,7,2
{82} 21,null,7,3
{83} 28,null,7,4
{84} 35,null,7,5
{85} 8,8,8,1
{86} 16,16,8,2
{87} 24,24,8,3
{88} 32,32,8,4
{89} 40,40,8,5
{90} 9,null,9,1
{91} 18,null,9,2
{92} 27,null,9,3
{93} 36,null,9,4
{94} 45,null,9,5
{95} 10,10,10,1
{96} 20,20,10,2
{97} 30,30,10,3
{98} 40,40,10,4
{99} 50,50,10,5
{100} 6,null,1,6
{101} 7,null,1,7
{102} 8,null,1,8
{103} 9,null,1,9
{104} 10,null,1,10
{105} 12,12,2,6
{106} 14,14,2,7
{107} 16,16,2,8
{108} 18,18,2,9
{109} 20,20,2,10
{110} 18,null,3,6
{111} 21,null,3,7
{112} 24,null,3,8
{113} 27,null,3,9
{114} 30,null,3,10
{115} 24,24,4,6
{116} 28,28,4,7
{117} 32,32,4,8
{118} 36,36,4,9
{119} 40,40,4,10
{120} 30,null,5,6
{121} 35,null,5,7
{122} 40,null,5,8
{123} 45,null,5,9
{124} 50,null,5,10
{125} 36,36,6,6
{126} 42,42,6,7
{127} 48,48,6,8
{128} 54,54,6,9
{129} 60,60,6,10
{130} 42,null,7,6
{131} 49,null,7,7
{132} 56,null,7,8
{133} 63,null,7,9
{134} 70,null,7,10
{135} 48,48,8,6
{136} 56,56,8,7
{137} 64,64,8,8
{138} 72,72,8,9
{139} 80,80,8,10
{140} 54,null,9,6
{141} 63,null,9,7
{142} 72,null,9,8
{143} 81,null,9,9
{144} 90,null,9,10
{145} 60,60,10,6
{146} 70,70,10,7
{147} 80,80,10,8
{148} 90,90,10,9
{149} 100,100,10,10
{150} 66,null,11,6
{151} 77,null,11,7
{152} 88,null,11,8
{153} 99,null,11,9
{154} 110,null,11,10
{155} 72,72,12,6
{156} 84,84,12,7
{157} 96,96,12,8
{158} 108,108,12,9
{159} 120,120,12,10
{160} 78,null,13,6
{161} 91,null,13,7
{162} 104,null,13,8
{163} 117,null,13,9
{164} 130,null,13,10
{165} 84,84,14,6
{166} 98,98,14,7
{167} 112,112,14,8
{168} 126,126,14,9
{169} 140,140,14,10
{170} 90,null,15,6
{171} 105,null,15,7
{172} 120,null,15,8
{173} 135,null,15,9
{174} 150,null,15,10
{175} 96,96,16,6
{176} 112,112,16,7
{177} 128,128,16,8
{178} 144,144,16,9
{179} 160,160,16,10
{180} 102,null,17,6
{181} 119,null,17,7
{182} 136,null,17,8
{183} 153,null,17,9
{184} 170,null,17,10
{185} 108,108,18,6
{186} 126,126,18,7
{187} 144,144,18,8
{188} 162,162,18,9
{189} 180,180,18,10
{190} 114,null,19,6
{191} 133,null,19,7
{192} 152,null,19,8
{193} 171,null,19,9
{194} 190,null,19,10
{195} 120,120,20,6
{196} 140,140,20,7
{197} 160,160,20,8
{198} 180,180,20,9
{199} 200,200,20,10
{200} 21,null,21,1
{201} 42,null,21,2
{202} 63,null,21,3
{203} 84,null,21,4
{204} 105,null,21,5
{205} 22,22,22,1
{206} 44,44,22,2
{207} 66,66,22,3
{208} 88,88,22,4
{209} 110,110,22,5
{210} 23,null,23,1
{211} 46,null,23,2
{212} 69,null,23,3
{213} 92,null,23,4
{214} 115,null,23,5
{215} 24,24,24,1
{216} 48,48,24,2
{217} 72,72,24,3
{218} 96,96,24,4
{219} 120,120,24,5
{220} 25,null,25,1
{221} 50,null,25,2
{222} 75,null,25,3
{223} 100,null,25,4
{224} 125,null,25,5
{225} 26,26,26,1
{226} 52,52,26,2
{227} 78,78,26,3
{228} 104,104,26,4
{229} 130,130,26,5
{230} 27,null,27,1
{231} 54,null,27,2
{232} 81,null,27,3
{233} 108,null,27,4
{234} 135,null,27,5
{235} 28,28,28,1
{236} 56,56,28,2
{237} 84,84,28,3
{238} 112,112,28,4
{239} 140,140,28,5
{240} 29,null,29,1
{241} 58,null,29,2
{242} 87,null,29,3
{243} 116,null,29,4
{244} 145,null,29,5
{245} 30,30,30,1
{246} 60,60,30,2
{247} 90,90,30,3
{248} 120,120,30,4
{249} 150,150,30,5
{250} 126,null,21,6
{251} 147,null,21,7
{252} 168,null,21,8
{253} 189,null,21,9
{254} 210,null,21,10
{255} 132,132,22,6
{256} 154,154,22,7
{257} 176,176,22,8
{258} 198,198,22,9
{259} 220,220,22,10
{260} 138,null,23,6
{261} 161,null,23,7
{262} 184,null,23,8
{263} 207,null,23,9
{264} 230,null,23,10
{265} 144,144,24,6
{266} 168,168,24,7
{267} 192,192,24,8
{268} 216,216,24,9
{269} 240,240,24,10
{270} 150,null,25,6
{271} 175,null,25,7
{272} 200,null,25,8
{273} 225,null,25,9
{274} 250,null,25,10
{275} 156,156,26,6
{276} 182,182,26,7
{277} 208,208,26,8
{278} 234,234,26,9
{279} 260,260,26,10
{280} 162,null,27,6
{281} 189,null,27,7
{282} 216,null,27,8
{283} 243,null,27,9
{284} 270,null,27,10
{285} 168,168,28,6
{286} 196,196,28,7
{287} 224,224,28,8
{288} 252,252,28,9
{289} 280,280,28,10
{290} 174,null,29,6
{291} 203,null,29,7
{292} 232,null,29,8
{293} 261,null,29,9
{294} 290,null,29,10
{295} 180,180,30,6
{296} 210,210,30,7
{297} 240,240,30,8
{298} 270,270,30,9
{299} 300,300,30,10
{chunk_no,dest_instance_id,source_instance_id} val
{j} x,y,i
{0} true,null,5
{1} true,true,6
{2} true,null,7
{3} true,true,8
{4} true,null,13
{5} true,true,14
{6} true,null,15
{7} true,true,16
{8} true,null,17
{9} true,true,18
{10} true,null,19
{11} true,true,20
{12} true,null,1
{13} true,true,2
{14} true,null,3
{15} true,true,4
{16} true,null,9
{17} true,true,10
{18} true,null,11
{19} true,true,12
{chunk_no,dest_instance_id,source_instance_id} val
{j} x,y,i
{0} 0.5,null,5
{1} 0.6,0.6,6
{2} 0.7,null,7
{3} 0.8,0.8,8
{4} 1.3,null,13
{5} 1.4,1.4,14
{6} 1.5,null,15
{7} 1.6,1.6,16
{8} 1.7,null,17
{9} 1.8,1.8,18
{10} 1.9,null,19
{11} 2,2,20
{12} 0.1,null,1
{13} 0.2,0.2,2
{14} 0.3,null,3
{15} 0.4,0.4,4
{16} 0.9,null,9
{17} 1,1,10
{18} 1.1,null,11
{19} 1.2,1.2,12
{chunk_no,dest_instance_id,source_instance_id} val
{j} x,y,i
{0} 0.5,null,5
{1} 0.6,0.6,6
{2} 0.7,null,7
{3} 0.8,0.8,8
{4} 1.3,null,13
{5} 1.4,1.4,14
{6} 1.5,null,15
{7} 1.6,1.6,16
{8} 1.7,null,17
{9} 1.8,1.8,18
{10} 1.9,null,19
{11} 2,2,20
{12} 0.1,null,1
{13} 0.2,0.2,2
{14} 0.3,null,3
{15} 0.4,0.4,4
{16} 0.9,null,9
{17} 1,1,10
{18} 1.1,null,11
{19} 1.2,1.2,12
{chunk_no,dest_instance_id,source_instance_id} val
{j} x,y,i
{0} 5,null,5
{1} 6,6,6
{2} 7,null,7
{3} 8,8,8
{4} 13,null,13
{5} 14,14,14
{6} 15,null,15
{7} 16,16,16
{8} 17,null,17
{9} 18,18,18
{10} 19,null,19
{11} 20,20,20
{12} 1,null,1
{13} 2,2,2
{14} 3,null,3
{15} 4,4,4
{16} 9,null,9
{17} 10,10,10
{18} 11,null,11
{19} 12,12,12
{chunk_no,dest_instance_id,source_instance_id} val
{j} x,y,z,i
{0} 5,null,null,5
{1} 6,0.6,'6',6
{2} 7,null,null,7
{3} 8,0.8,'8',8
{4} 13,null,null,13
{5} 14,1.4,'14',14
{6} 15,null,null,15
{7} 16,1.6,'16',16
{8} 17,null,null,17
{9} 18,1.8,'18',18
{10} 19,null,null,19
{11} 20,2,'20',20
{12} 1,null,null,1
{13} 2,0.2,'2',2
{14} 3,null,null,3
{15} 4,0.4,'4',4
{16} 9,null,null,9
{17} 10,1,'10',10
{18} 11,null,null,11
{19} 12,1.2,'12',12
{chunk_no,dest_instance_id,source_instance_id} val
{j} x,i
{0} 5,5
{1} 15,15
{2} 16,16
{3} 17,17
{4} 18,18
{5} 19,19
{6} 20,20
{7} 1,1
{8} 2,2
{9} 3,3
{10} 4,4
UserException in file: src/smgr/io/TemplateParser.cpp function: parse
Error id: scidb::SCIDB_SE_EXECUTION::SCIDB_LE_ATTRIBUTES_MISMATCH
Error description: Error during query execution. Attributes mismatch.
UserException in file: src/smgr/io/TemplateParser.cpp function: parse
Error id: scidb::SCIDB_SE_EXECUTION::SCIDB_LE_ATTRIBUTES_MISMATCH
Error description: Error during query execution. Attributes mismatch.
{chunk_no,dest_instance_id,source_instance_id} val
